### ISSUE TITLE
Allow `torch.fx` tracing on TorchScript models

### ISSUE THREAD

**Issue by alanhdu** [pytorch/pytorch#76181]
### 🚀 The feature, motivation and pitch

It would be really nice if we could use `torch.fx.symbolic_trace` on a TorchScript model. We use TorchScript extensively to save models since it's a hermetically sealed artifact that no longer requires a Python dependency. We'd like to be able to then `torch.fx.symbolic_trace` these TorchScript models so we can use the Graph FX quantization API on pre-trained models -- right now, if you try to trace a TorchScript model you get an error like:

```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [47], in <module>
      1 original = nn.Linear(1, 1)
      2 scripted = torch.jit.script(original)
----> 3 torch.fx.symbolic_trace(scripted)

File /opt/conda/envs/ctrldev/lib/python3.8/site-packages/torch/fx/_symbolic_trace.py:907, in symbolic_trace(root, concrete_args, enable_cpatching)
    858 """
    859 Symbolic tracing API
    860 
   (...)
    904     GraphModule: a Module created from the recorded operations from ``root``.
    905 """
    906 tracer = Tracer(enable_cpatching=enable_cpatching)
--> 907 graph = tracer.trace(root, concrete_args)
    908 name = root.__class__.__name__ if isinstance(root, torch.nn.Module) else root.__name__
    909 return GraphModule(tracer.root, graph, name)

File /opt/conda/envs/ctrldev/lib/python3.8/site-packages/torch/fx/_symbolic_trace.py:559, in Tracer.trace(self, root, concrete_args)
    557 if isinstance(root, torch.nn.Module):
    558     self.root = root
--> 559     fn = type(root).forward
    560     self.submodule_paths = {mod: name for name, mod in root.named_modules()}
    561 else:

File /opt/conda/envs/ctrldev/lib/python3.8/site-packages/torch/jit/_script.py:298, in _CachedForward.__get__(self, obj, cls)
    297 def __get__(self, obj, cls):
--> 298     return self.__getattr__("forward")

AttributeError: '_CachedForward' object has no attribute '__getattr__'
```

This would be super convenient because it would decouple the code that produced the original pre-trained model and the code used for quantization.

### Alternatives

Ultimately, what we want is to be able to quantize a model without having access to the original code that generated the model (so no `pickle`, for instance). AFAICT, TorchScript is the best way to serialize a model without needing the original Python dependency, but if there's an alternative serialization format that plays nice with `torch.fx` then I think we'd be happy to use that instead (e.g. is there a way to serialize a `torch.fx.Graph` in such a hermetically sealed way?).

**Comment by saurabh-shandilya** [comment]
I agree. I am not sure how do I use torch.fx quanization API to quantize a torchscript model. Torchscript models are pretty self contained and widely used for sharing/exchange. jit_quantize currently works well for that. But torch.fxcan  be a much better replacement for that since it allows more powerful transforms, but there is no entry door into this that I am aware of. Only outgoing no incoming!! :-(

**Comment by joseph-zhong** [comment]
+1 This is highly blocking for our project which requires FX-Quantization from already serialized TorchScript objects 

Is the best practice to quantize and serialize afterwards? Does that mean if we don't know a priori whether a model can or can't be quantized, we should just try-except the quantization and then serialize afterwards?

ex. workflow

```python
# train model ...
model = ...

# try to quantize
try:    
    model.quantize(...)
    ... 
except Exception as e:
    logger.error(f"Failed to quantize with error '{e}', may contain unsupported ops!")
finally:
    # TorchScript serialize
    compiled_model = torch.jit.script(model) 
    torch.jit.save(compiled_model, 'compiled_model.pt')
```

**Comment by ArshadIram** [comment]
I am also getting this issue. It was working yesterday but today I am not able to import model_zoo.

---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
[<ipython-input-2-500ff0b58130>](https://localhost:8080/#) in <module>
     11 
     12 # import some common detectron2 utilities
---> 13 from detectron2 import model_zoo
     14 from detectron2.engine import DefaultPredictor
     15 from detectron2.config import get_cfg

9 frames
[/usr/local/lib/python3.7/dist-packages/detectron2/utils/tracing.py](https://localhost:8080/#) in <module>
      3 import torch
      4 from torch.fx._symbolic_trace import _orig_module_call
----> 5 from torch.fx._symbolic_trace import is_fx_tracing as is_fx_tracing_current
      6 
      7 from detectron2.utils.env import TORCH_VERSION

ImportError: cannot import name 'is_fx_tracing' from 'torch.fx._symbolic_trace' (/usr/local/lib/python3.7/dist-packages/torch/fx/_symbolic_trace.py)

---------------------------------------------------------------------------
NOTE: If your import is failing due to a missing package, you can
manually install dependencies using either !pip or !apt.

To view examples of installing some common dependencies, click the
"Open Examples" button below.

**Comment by mert-kurttutan** [comment]
I agree that torch.fx on script compiled would be super convenient.

I think this issue is related to the way models are traced by torch.fx. 
During tracing, it overwrites `__torch_function__` method, which apparently does not work with compiled models. See this closely related issue, [link to issue](https://github.com/pytorch/pytorch/issues/57739)

**Comment by lopusz** [comment]
+1 for this issue. 

I find torch.fx quite amazing. However, lack of support for script models is quite a limitation. 

From my viewpoint this is extremely practical use case: 

1. load scripted model, 
2. torch.fx.symbolic_trace  it, 
3. do some fx transformations and fine tuning (e.g. fx.quantization)
4. script and save back improved model

torch.fx.GraphModule -> script works fine, script -> torch.fx.GraphModule is not supported.
 
Are there any plans to support it?

Indeed, as @alanhdu  mentioned an alternative would be some serialization format (other than script) which allows to load models without installing dependencies. Perhaps some new method is in the making (among other 2.0+ improvements)?

**Comment by lopusz** [comment]
After a bit of browsing the issues, I believe this is a duplicate of #61464.

Closely related to #57124.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.

**Comment by joseph-zhong** [comment]
thanks Jane!
